Normalisation refactor

ZedThree opened the pull request:
Refactor of #85

Replaces `NOCOS` with `NormalisationConvention`. The different codes' conventions are described as modifications to the base `pyrokinetics` convention:

```python
NORMALISATION_CONVENTIONS = {
    "pyrokinetics": NormalisationConvention("pyrokinetics"),
    "cgyro": NormalisationConvention("cgyro", bref_type="Bunit"),
    "gs2": NormalisationConvention("gs2", vref_multiplier=np.sqrt(2)),
    "gene": NormalisationConvention("gene", lref_type="major_radius"),
}
```

I switched these to named conventions as being a bit more user-friendly. It's now a bit easier to add new ones, so I think it would be fine to have e.g. `stella` be the same as `gs2` in practice.

I've renamed `LocalNorm` to `Normalisation`, but if pyrokinetics gains the ability to handle global codes, the original name might be better.

`Normalisation.from_kinetics` and `from_local_geometry` are class methods that return a new instance instead of modifying the existing one. I'm not completely happy about this, because really we need both at the same time. I think this is sort of just a symptom of how we can create a `pyro` object piecemeal from kinetics and geometry.

bpatel2107 committed:
Added a LocalNorm Object that stores the normalisations and has several different options for different codes

bpatel2107 committed:
Uncomment out group of tests

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Update GENE "default" vref_multiplier in NOCOs

bpatel2107 committed:
Merge remote-tracking branch 'origin/enhancement/normalisation_options' into enhancement/normalisation_options

bpatel2107 committed:
Flake8 fixes

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Update relative imports

bpatel2107 committed:
Revert relative import

bpatel2107 committed:
Remove unused variable

bpatel2107 committed:
Made LocalNorm module

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Merge unstable into normalisation branch

bpatel2107 committed:
Merge branch 'enhancement/normalisation_options' of github.com:bpatel2107/pyrokinetics into enhancement/normalisation_options

 Conflicts:
	pyrokinetics/gk_code/GKCodeGENE.py

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Delete GKCodeGENE.py

bpatel2107 committed:
Delete GKCodeGS2.py

bpatel2107 committed:
Remove repeated line

bpatel2107 committed:
Fix error in CGYRO Input set

bpatel2107 committed:
Merge remote-tracking branch 'origin/enhancement/normalisation_options' into enhancement/normalisation_options

bpatel2107 committed:
Fix flake8 errors

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Add local_norm in update_gk_code

bpatel2107 committed:
Merge remote-tracking branch 'origin/enhancement/normalisation_options' into enhancement/normalisation_options

ZedThree committed:
Convert NOCOS to simpler dataclass

ZedThree committed:
Small refactor of `LocalNorm` internals

ZedThree committed:
Add `__repr__` for `LocalNorm`

ZedThree committed:
Change `LocalNorm.from_*` methods to constructors

ZedThree committed:
Simplify some uses of `local_norm` in GK inputs

ZedThree committed:
Pass normalisation into `LocalSpecies.from_kinetics`

ZedThree committed:
Rename `LocalNorm` -> `Normalisation`

ZedThree committed:
Use plain attribute instead of property for normalisation

ZedThree committed:
Fix bad merge: `LocalSpecies.nspec` is now a property

ZedThree committed:
Make roundtrip test assert clearer

ZedThree committed:
Make `local_norm` optional argument to `GKInput.set`

ZedThree committed:
Create normalisation before local_species

ZedThree committed:
Add some type annotation to Normalisation

ZedThree committed:
Make `psi_n` a required argument for `LocalSpecies.from_kinetics`

ZedThree committed:
[skip ci] Apply black changes

ZedThree commented on pyrokinetics/local_species.py:
```diff
@@ -61,33 +60,20 @@ def from_dict(self, species_dict, **kwargs):
         return local_species
 
-    def from_kinetics(self, kinetics, psi_n=None, lref=None):
+    def from_kinetics(self, kinetics, psi_n, local_norm):
```
I'm not actually convinced that passing in `local_norm` is correct here -- previously it was generated to be consistent with `kinetics`, although `lref` had to be passed in.

Maybe it should be optional, and if `None`, then it uses `kinetics` for the normalisation?

bpatel2107 replied:
`lref` isn't stored in `kinetics` so how else would we get it? We could directly pass `lref` by doing

```python
local_species.from_kinetics(kinetics, psi_n=psi_n, lref=local_norm.lref)
```

ZedThree commented on pyrokinetics/pyro.py:
```diff
@@ -1338,12 +1322,20 @@ def load_local_species(self, psi_n: float, a_minor: Optional[float] = None) -> N
+            psi_n=psi_n,
+            convention=self.local_norm.nocos.name,
+            lref=a_minor,
+            bref=self.local_norm.bref,
```
Here we need to reuse `bref` from that calculated during `load_geometry` above. Looking at this with fresh eyes, this does put a dependency on the order in which `load_geometry` and `load_local_species` should be called.

bpatel2107 replied:
Yes this is true. We should add a check to make sure `local_geometry` is in before `local_species` is loaded

bpatel2107 commented:
We'll need to add in TGLF as well as the outputs.

I'm also not sure about how to actually implement these changes when writing to a file. Do we store everything internally using the pyrokinetics convention and then convert when writing the files or do we store everything in the requested convention and then write the data as is.

The latter requires the geometry/local_species to be re-calculated. Like right now these changes have no impact on `local_geometry.Rmaj` which technically should be `Rmajor / Lref`

bpatel2107 commented:
> 
> I'm also not sure about how to actually implement these changes when writing to a file. Do we store everything internally using the pyrokinetics convention and then convert when writing the files or do we store everything in the requested convention and then write the data as is.
> 

@ZedThree  @d7919 thoughts?

d7919 commented:
I'm probably not familiar enough with the current code structure to make an informed comment, but my uninformed feeling is that always storing in pyrokinetics and then converting to the requested convention on write might be the cleanest.

bpatel2107 commented:
I think I agree. This will require adding the normalisations to the readers/writers but I think it is worth it in the long run.

Only tricky detail would be setting values becomes awkward as you need to do it in pyro conventions and then let pyro change it for you. E.g setting beta in CGYRO requires knowing to multiply by `(B0/Bunit)**2`

d7919 commented:
If you're using a setter, rather than directly accessing variables I suppose it may be possible to add an optional 'convention' input to specify what convention you're using when you pass something?

bpatel2107 commented:
Would that be quite a large refactor?

d7919 commented:
I would guess it could be, although maybe someone knows some magic to achieve this nicely

bpatel2107 commented:
@ZedThree if we want to store everything internally with one normalisation (pyrokinetics by default), do you know of a clever way of reading/writing the data and always applying the correct normalisation? 

Or would we need to go through each variable and apply the normalisation (which would could be done using the `NormalisationConventions`) Something like 

```python
gs2['rmaj'] = miller.rmaj * NormalisationConvention.convert("pyrokinetics", "gs2", "lref")
```

ZedThree commented:
It's worth looking at [pint](https://pint.readthedocs.io/en/stable/index.html) for handling units. Then we can rely on that to do the conversions for us. There's a pint plugin for xarray too.

ZedThree committed:
Merge branch 'unstable' into ph-normalisation

* unstable: (87 commits)
  Don't write modified test gs2 input file to disk
  Add tests to check reading GS2 nonlinear input files
  Add tests for more GS2 input file variations
  Allow `ngauss` to be unset in GS2 input files
  Read GS2 box grids without jtwist explicitly set
  Don't hide deeper errors when reading GS2 grid from input file
  Add some tests for GS2 input file variations
  Fix for checking `wstar_units` set in nonlinear GS2 inputs
  Set local_geometry attr as Miller for Circular types
  Update circular geometry Rmaj, rho
  [skip ci] Apply black changes
  Write out correct omn,omt in GENE
  Handle cases when omn uses Lref!=aminor
  Correct definitions of rho and Rmaj in GKInputGENE.py
  [skip ci] Apply black changes
  Defined kx, ky and theta0 in GS2 single_grid
  Drop dimensions when storing eigfunc in pyroscan
  [skip ci] Apply black changes
  GENE circular geometry read capability added
  Fix "not in" in if statement
  ...

ZedThree committed:
Fix new normalisation for TGLF

ZedThree commented:
I've spent today playing about with `pint`, and I think I can get it working like this:

```python
# Add units to Rmaj
Rmaj = pyro.local_geometry.Rmaj * ureg.metre

# Convert to GENE normalisation:
print(Rmaj.to("gene_lref"))
# 1.0 gene_lref

# In a given context, we can drop the convention name from the reference values
with ureg.context("gs2"):
    print(Rmaj.to("lref"))
# 1.233172498241725 lref
```

With this approach, we always create all the normalisation conventions, and then it doesn't matter what convention we store values in, we can always trivially convert to some another convention:

```python
Rmaj.to("gs2_lref").to("pyrokinetics_lref").to("gkdb_lref").to("gene_lref")
```
(although from my testing this does incur some rounding errors)

What I haven't been able to get working nicely is where something needs to be normalised to some combination of reference values, doing that without repeating the normalisation. That is, this works:

```python
with ureg.context("gene"):
    # Normalised to GS2
    nu = 5.677 * ureg.gs2_lref / ureg.gs2_vref
    print(nu.to("lref / vref"))
# 3.255217905458966 lref / vref
```

But getting something like `nu.to("gs2")` seems much more difficult.

We would need to make sure things were in the correct units first, before we could use this to normalise them, but I'm not sure how far we'd need to go. It would probably be nice if _everything_ had units, then we could also handle converting growth rates into say, Alfvén wave transit times or whatever.

bpatel2107 commented:
I think we can add units to everything with relative ease, just will take time.

Do you think we could make a light wrapper around pint that handles cases like `nu.to("GS2")` that could handle these cases. Most inputs are going to be a combination of reference values so it would be good in the long run

ZedThree committed:
WIP: Start adding functions for using `pint` units

ZedThree committed:
Always enable Boltzmann context, to convert eV to K

ZedThree committed:
Add base unit of time for each normalisation convention

ZedThree committed:
Add proxy objects which nicer names to unique pint units

ZedThree committed:
Add method to `pint.Quantity` for converting between systems

ZedThree committed:
Give each `Pyro` instance a unique name

ZedThree committed:
Rename normalisation current convention to work with to_base_units

ZedThree committed:
Add magic method for copying `SimulationNormalisation`

ZedThree committed:
Enable easier conversion between different normalisation conventions

ZedThree committed:
Split method for setting lref, bref into two

Not entirely convinced this is necessary

ZedThree committed:
Make `beta` a property rather than a defined unit

ZedThree committed:
Switch normalisation implementation over to unit-based version

ZedThree commented:
I had to explore a few dead-ends and investigate several libraries, but I now have something that mostly works, it just has lots of knock-on effects that need to be fixed. But we can now have things like:

```python
In [3]: pyro.local_species.electron.nu
Out[3]: 0.050767795211818026 <Unit('vref_pyrokinetics_run0000 / lref_pyrokinetics_run0000')>

In [4]: pyro.local_species.electron.nu.to(pyro.norms.gs2)
Out[4]: 0.03589825226016644 <Unit('vref_gs2_run0000 / lref_gs2_run0000')>

In [5]: pyro.local_species.electron.nu.to_base_units()
Out[5]: 20311.324112597606 <Unit('1 / second')>

In [6]: pyro2.local_species.electron.nu.to(pyro.norms)
Out[6]: 0.0183350430079753 <Unit('vref_pyrokinetics_run0000 / lref_pyrokinetics_run0000')>

In [7]: pyro.local_species.electron.nu.to(pyro2.norms)
Out[8]: 0.03942599128526476 <Unit('vref_pyrokinetics_run0001 / lref_pyrokinetics_run0001')>
```
(where `pyro` and `pyro2` are different simulations)

We could in principle define other units and conventions, like the cyclotron frequency and Larmor radius.

Each `Pyro` object gets a unique name which is used in turn to give each combination of simulation and convention a unique name for each unit. The stem can be set manually, taken from one of the input files, or the generic fallback "run". We create these units for all the conventions at once through a top-level `SimulationNormalisation` object, owned by each `Pyro` instance, which has a `ConventionNormalisation` for each convention. The `SimulationNormalisation` object has `lref`, `bref`, etc members which point to the current `default_convention`'s specific units. So `pyro.norms.lref` points to `pyro.norms.pyrokinetics.lref`, which points to the actual `lref_pyrokinetics_<run_name>`, by default.

It's still failing a couple of tests, mostly issues around PyroScan and copying these normalisation objects. There's also lots of places that now need the normalisation and units adding. Hopefully I'll be able to finish it off next week.

Some important caveats and downsides:

- I've picked `pint`, but there is also `astropy.units` and `unyt`. `plasmapy`, which would be really interesting/useful to incorporate, uses `astropy.units`. Unfortunately, that doesn't play nice with `xarray`, whereas `pint` does. If we do end up using `plasmapy`, we can write something to wrap their functions and convert the unit systems
- It's not possible to convert dimensionless quantities like `beta` between normalisations. I _think_ it would be possible to re-jig things to make this possible, by giving it a new dimension like `[beta_gs2]` and having explicit conversion functions. Currently, the different normalisations have their own `beta` value, like `norms.gs2.beta` and `norms.cgyro.beta`, that need to be accessed explicitly.
- It's not currently possible to convert between different codes without having these normalisation structures created from the geometry/kinetic data. I think this would be possible in the future, similar to the dimensionless quantities, by using parameterised contexts, but this has some severe footguns
- When a variable becomes a `Quantity` (i.e. has some units), they might need to be converted back to a normal number with `value.magnitude` or `value.m` for short

Still needs some specific tests for the normalisation stuff, and some proper documentation on how this works. It might also be nice to have something like `pyro.norm_to(pyro.norms.gs2)` or similar, to convert a whole simulation to a different convention, but that can come later if needed.

@bpatel2107 Do you want to have a play with this version and see if if breaks in any unexpected ways? It would also be great if we could add units to as many things as possible, but that can probably also be done piecemeal. E.g. adding units to the geometry and everything downstream from that.

ZedThree committed:
[skip ci] Apply black changes

ZedThree committed:
Remove some unused imports

ZedThree committed:
Define normalised current in terms of charge, length, velocity

ZedThree committed:
Add some docs and comments for normalisations

ZedThree committed:
Fix: Length scales are dimensionless

ZedThree committed:
Enable comparing Quantities and values in tests

ZedThree committed:
Remove ignore keys in test, now that reference values are in norms

ZedThree committed:
Add contexts for converting simulations without physical references

ZedThree committed:
Add tests for normalisations

ZedThree committed:
Simplify converting simulation quantities to physical references

ZedThree committed:
Fix use of deprecated `gk_type` in tests

ZedThree committed:
Fix normalisations for mixed simulation units

ZedThree committed:
First pass at parameterising test_roundtrip

ZedThree committed:
Add units to GK inputs

ZedThree committed:
Refactor normalisations to reuse units between conventions

Instead of unique dimensions for each combination of simulation and
unit, instead have unique dimensions for each simulation unit only,
and then have transformations between simulation and physical
units. Also try to define simulation units in terms of other ones --
for instance, `vref_most_probable = sqrt(2) * vref_nrl` -- which
allows us to convert between them without extra faff.

ZedThree committed:
Remove unused imports

ZedThree committed:
Pin numpy version below 1.21.0

That version drops support for Python 3.7

ZedThree committed:
Drop support for python 3.7

ZedThree commented:
Ok, I have spent way too long on this, but it does now work and pass all the tests. There's probably more places to add units, but the basics are covered. This is ready for review!

There's a bunch of docs, but basically, there's now a set of base "simulation units" that are defined in terms of fictional dimensions, such as `lref_minor_radius = [lref]`, or other simulation units, such as `lref_major_radius = <aspect ratio> lref_minor_radius`. Then there are a set of conventions that collate the appropriate units, e.g. GS2 has `lref = lref_minor_radius`, GENE has `lref = lref_major_radius`.

We can then provide physical reference values to get physical units that have unique names per simulation, and are defined in SI units, so `lref_minor_radius_run1234 = 1.5 metres`. We update each convention's units for a given simulation to the physical unit instead of the simulation unit.

So now we can use the conventions to convert units:

```python
growth_rate_gene = growth_rate_gs2.to(norms.gene)
```

It could definitely still use some polish in places, and I'm sure this is not the most elegant solution, but it's a decent start at least. 

Better names for anything/everything also welcome. For example, the velocity units are named according to the methods in the [plasmapy docs](https://docs.plasmapy.org/en/stable/api/plasmapy.formulary.speeds.thermal_speed.html#thermal-speed-notes), `vref_nrl` and `vref_most_probable` -- which are not really very good names!

---

The latest version of numpy and pint don't support Python 3.7, so rather than fight the CI, I've just dropped it...

bpatel2107 commented on pyrokinetics/gk_code/GKInputGENE.py:
```diff
@@ -120,8 +129,6 @@ def get_local_geometry_miller(self) -> LocalGeometryMiller:
 
```
Here we are reading in the data and working out the pyro normalisation. This should be fine when we are writing out new files and running new simulations. But if we wanted to load in an existing simulations that has `lref=major_radius` (which corresponds to `self.data["geometry"]["major_r"] = 1.0` then pyro wouldn't know how to normalise the outputs. I think when we add in the normalisation for the outputs we need to check the input file to see which normalisation was used.

ZedThree replied:
I was a bit confused by this -- if I understand correctly, GENE input files always (sometimes?) have both `major_r` _and_ `minor_r` set, and then Pyrokinetics will renormalise it so that `major_r == 1.0`? I don't have access to the GENE manual -- are these in metres, or normalised units?

With this PR, reading in any GENE file will result in lengths being defined in terms of `lref_major_radius`, and writing back out to GENE (_should_) be idempotent (i.e. not change anything physically), and to convert to other codes we need either a) both `Rmaj` and `a_minor`, or b) the aspect ratio. In practice, we seem to always have `Rmaj` set and only need to supply `a_minor`.

We currently only set `Rmaj` in the Miller geometry from GENE input files, but maybe we should also set `a_minor`? This way, we'd always have the correct aspect ratio used in GENE and so wouldn't need to supply it. How many other codes include the aspect ratio? I believe GS2 can take it as an option (at least two different ways, I'm sure).

I did consider adding the following to `LocalGeometry`:

```diff
 
+    @property
+    def aspect_ratio(self) -> float:
+        return self.Rmaj / self.a_minor
+
+    @aspect_ratio.setter
+    def aspect_ratio(self, value: float):
+        # TODO: Does it make sense to change the aspect ratio if both
+        # R and r are already set?
+        if hasattr(self, "Rmaj"):
+            self.a_minor = self.Rmaj / value
+        elif hasattr(self, "a_minor"):
+            self.Rmaj = self.a_minor * value
+        else:
+            raise RuntimeError(
+                "Cannot set aspect_ratio: missing both minor and major radius"
+            )
+
+    @property
+    def inverse_aspect_ratio(self) -> None:
+        return 1.0 / self.aspect_ratio
+
+    @inverse_aspect_ratio.setter
+    def inverse_aspect_ratio(self, value: float) -> None:
+        self.aspect_ratio = 1.0 / value
+
     # TODO replace this with an abstract classmethod
```

bpatel2107 commented:
Overall it looks pretty good so far. We need to add in the output normalisations too. Do you want to leave that for a separate PR or include it in this one?

ZedThree commented:
I think it would be better to get this in so people can play around with it and find the bugs, and then add more units after.

To convert between simulation and physical units, we need to have a pint `Context` with an explicit transformation, and pint won't transform between mixed units. For example, there's a context (automatically enabled in this PR) to convert between `eV` and `K`, but pint can't convert `eV / m` to `K / m`. So I've add to explicitly add some transformations for what I think are the most common mixed units: `[vref] / [lref]` and `[nref] * [tref]` for instance. If people regularly use other mixed units, we'll need to add them.

ZedThree commented:
In that case, I think we'd still need to keep `nref` and `tref` in there, as otherwise you wouldn't be able to tell the difference between the betas for simulations that did or didn't have nref/tref. For instance, you'd read in a GK file, get one dimensionless value for beta, then add the physical reference values and get a different, but still dimensionless, beta value.

I think I can probably hack this for now so that it works like you want, but with ugly units.

bpatel2107 commented:
Technically if beta is already defined then `bref` should be calculated when setting `tref` and `nref`. Maybe we can add a check to see if the value of beta is consistent with this or not.

ZedThree commented:
Ah, good point. Hmm, that probably complicates things a bit, as pint doesn't really handle dimensionless quantities very well.

I rather suspect that means we'll need a `[beta]` dimension. Not sure about transformations to/from `bref`, hopefully just the consistency check as you suggest will do

bpatel2107 commented:
I think we can have a beta scale factor out the front that gets set such that when we load in the reference values the actual value used in simulations doesn't change and a warning is printed if it is not 1

bpatel2107 commented:
There will be cases when people want to do a beta scan so we need some way to modify the beta used in a simulation. 

We can figure out which species is the reference species by their values. i.e. if `local_species.electron.temp = 1.0` then tref = te and so on.

This is why a scale factor might be best as it separates the reference beta which we can define as `beta_ref = 8*pi*nref*tref / bref**2` and in a simulation we do `gk_code.beta = beta_multiplier * beta_ref` if it is defined.

This way `beta_ref` becomes a unit and the GK `beta` is just relative to that value.

I do like the idea of a warning when not consistent though

ZedThree commented:
Sure, we definitely want to support people running beta scans.

Where does all this live though? I'm not sure about putting `beta_multiplier` into the norm classes, as then they're not purely reference values and units. Having it elsewhere though will make it really difficult to convert beta between conventions -- we can't make it a simple unit, because it's dimensionless and therefore impossible to convert automatically. We can do it manually, of course, provided we do know which conventions we converting to/from.

It might be a bit easier in some ways if we did give beta its own dimension, but working out all the different conversions becomes difficult then too, because pint is very awkward for converting dimensions. We'd also need a different dimension for each combination of `(nref, tref, bref)`.

I'll see if I can get it work with separate dimensions, as that will give us the most flexibility in terms of how to store the actual simulation value.

bpatel2107 commented:
To be honest its not simple. Right now beta sneakily lives inside the `local_geometry` via `B0` but its not the best place for it.  

There may be need to have a new object that stores these kind of parameters that need reference values. For now we can think about storing `beta_multiplier` in the base pyro or the numerics object?

ZedThree commented:
Yes, maybe in numerics might make most sense?

I _think_ I can make it work by defining a dimension and setting up the "other" conventions as units, basically the same as I've done for all the other reference values. Then we probably only need to hardcode changing the units from dimensionless to `[beta_ref]`

ZedThree commented:
I think the `beta_multiplier` is a bit too complicated. Here's what I think are all the possible situations, and what variables _could_ be defined in each case:

| | `beta_ref` | `1/B0**2` | `beta` | `beta_multplier` |
|--|--|--|--|--|
| geometry + kinetic | :heavy_check_mark: | :heavy_check_mark: | :x: | :x: |
| GK + geometry + kinetic | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| GK + geometry | :x: | :heavy_check_mark: | :heavy_check_mark: | :white_check_mark: |
| GK + kinetic | :x: | :x: | :heavy_check_mark: | :x: |
| GK | :x: | :x: | :heavy_check_mark: | :x: |

where `beta` is the value actually used in the simulation.

We can always set `beta` for a GK simulation, even if it is done so arbitrarily. `beta_multiplier` can only really be set in the "full physics" case. I think it would be easiest to set `numerics.beta` from the GK simulation and warn if it's not consistent with `beta_ref`. If we have units like `beta_ee_B0`/`beta_ee_Bunit`, we can then automatically convert (or error!) to the correct convention.

Is `1/B0**2` really used very often as an approximation to `beta`? It feels very off to me: the units are wrong, it looks like it's saying that `n*T ~ 1/(2*mu0)`, and cannot be kept consistent between a GK simulation and `beta_ref`.

bpatel2107 commented:
The `1/B0**2` was not physically motivated and I am open to changing this. I set it to that as it was the closest thing I could come up with that was sort of physically relevant though I do see how it isn't great.

I am fine with setting the GK `beta` in the `numerics` and having this warning pop up. In the case where we do have `nref`, `tref` and `bref` and we use a template file for a GK input we should set `beta` to `beta_ref` by default.

ZedThree committed:
Document which normalisation convention each GK code uses

ZedThree committed:
Add units for beta

ZedThree commented:
Looking at how GKDB/IMAS handles it, they have `beta_reference` in `species_all`, which I think corresponds to our `LocalSpecies`.

I'm just going to stick it into `numerics` for now, and we should have a bigger project of renaming and moving things to better align with IMAS later.

ZedThree committed:
Merge branch 'unstable' into ph-normalisation

* unstable:
  Drop support for Python 3.7
  [skip ci] Apply black changes
  Fix GENE reader for missing `geometry::minor_r` input

bpatel2107 commented:
I think numerics is better as you can't get beta from the local species data alone. So you wouldn't be able to fully load in `LocalSpecies` from just a SCENE/TRANSP output.

This along with `rho_star` (which isn't stored yet) are the only terms I can think of that need both local species data and local geometry data so shouldn't really be stored in either one alone.

Once this is in we can discuss how handle it going forward

ZedThree committed:
Enable user-changeable `beta`

`beta` now stored in `Numerics`

When writing GK input files, `beta` defaults to value in `Numerics`,
and falls back to reference beta if `numerics.beta` is zero

ZedThree committed:
Add normalisation to built docs

ZedThree commented:
Now you can change `pyro.numerics.beta` however you like, and that's what will get written to file -- unless it's zero, in which case it will fallback to the reference beta value. Is that actually what we want though? How do we handle electrostatic cases? Set it to `None` maybe?

ZedThree committed:
Remove unused import

bpatel2107 commented:
I think if `pyro.numerics.beta =  0.0` then the GK input file should be 0. You can imagine doing a scan from 0 to 1 and you would want this behaviour.

If it is `None` then it should try and set it to the reference value and if it can't do that then set it to 0

ZedThree committed:
Only fallback to reference beta if `numerics.beta` is `None`

Allows user setting `beta` to zero

ZedThree committed:
Ensure we don't write possibly unconverted values or NaN to file

If we convert a value and it becomes NaN, that indicates we don't have
a physical unit for the equivalent simulation unit. So we should
error, to avoid passing on values in the wrong units.

Currently special cases zero, because this usually indicates a value
is unset -- perhaps we should make sure we use None instead?

ZedThree commented:
The last commit actually caught a bug: collision frequencies were not being converted properly. Unfortunately, this means we basically always need both geometry and kinetic files in order to convert GK files. This is because `nu ~ vref / lref`, and `lref` comes from the geometry, while `vref` comes from the kinetics. In principle we should be able to convert `nu` if we know the aspect ratio alone, but unfortunately pint is very annoying when it comes to mixed units and I can't make it work.

The workaround is to just zero the collision frequency for all species, or to convert to the correct units manually.

Similarly for `beta`, where we can read it in from a GK file, but cannot convert it between conventions without the geometry (for `bref`)

ZedThree committed:
Add context so vref/lref can be converted with only aspect ratio set

bpatel2107 commented:
Can you do something similar for `bref` to convert from B0 normalisation to Bunit normalisation?

ZedThree commented:
That already works! Although currently you need to call `set_bref` with a `LocalGeometry` -- would be trivial to add `B0` and `bunit_over_b0` arguments though

bpatel2107 commented:
Okay awesome! Is this good to merge now?

ZedThree commented:
Yep! It needs some real-world testing now, and then we can start to add units to the rest of the code
